src/lib.rs

```diff
@@ -1502,10 +1510,61 @@ impl rustc_serialize::Decodable for Url {
     }
 }
 
+#[cfg(feature="serde")]
+impl Url {
+    /// Serialize the URL efficiently, for use with IPC
+    pub fn serialize_efficient<S>(&self, serializer: &mut S) -> Result<(), S::Error> where S: serde::Serializer {
+        use serde::Serialize;
+        // Destructuring first lets us ensure that adding or removing fields forces this method
+        // to be updated
+        let Url { ref serialization, ref scheme_end,
+                  ref username_end, ref host_start,
+                  ref host_end, ref host, ref port,
+                  ref path_start, ref query_start,
+                  ref fragment_start} = *self;
+        (serialization, scheme_end, username_end,
+         host_start, host_end, host, port, path_start,
+         query_start, fragment_start).serialize(serializer)
+    }
+
+    /// Deserialize the URL efficiently, without re-parsing
+    ///
+    /// Deserializer must be known to contain the output of serialize_efficient
+    ///
+    /// Cannot cause memory unsafety if used incorrectly, but may cause security issues.
+    /// Only use with trusted input.
+    pub fn deserialize_efficient<D>(deserializer: &mut D)
+        -> Result<Url, D::Error>
+        where D: serde::Deserializer {
+        use serde::{Deserialize, Error};
+        let (serialization, scheme_end, username_end,
+         host_start, host_end, host, port, path_start,
+         query_start, fragment_start) = Deserialize::deserialize(deserializer)?;
+        let url = Url {
+            serialization: serialization,
+            scheme_end: scheme_end,
+            username_end: username_end,
+            host_start: host_start,
+            host_end: host_end,
+            host: host,
+            port: port,
+            path_start: path_start,
+            query_start: query_start,
+            fragment_start: fragment_start
+        };
+        #[cfg(debug_assertions)] {
```

Comment:
Without `debug_assertions`, type errors in that block won't be emitted.  Do `if cfg!(debug_assertions)` instead.